File: Sources/Heeler/Notifications/NotificationRegistrationCeremony.swift
import Foundation

/// The per-Host Notification Registration ceremony (#72, ADR 0008): mint or
/// reuse the Host's Notification Key, then write this device's entry into
/// the registration file the Host's notify hook reads. The Transport owns
/// where that file lives and how it is replaced atomically; this type owns
/// the read-merge-write and the custody split (key in the device Keychain and
/// Host registration file, token only in the Host registration file).
///
/// Every failure is surfaced: a thrown `NotificationRegistrationError` or
/// `TransportError` means notifications are not armed, never silently broken.
struct NotificationRegistrationCeremony: Sendable {
    let keys: NotificationKeyStore

    init(keys: NotificationKeyStore = NotificationKeyStore()) {
        self.keys = keys
    }

    /// Registers this device for Agent Notifications from one Host. The
    /// Notification Key is saved locally before the remote write: a write
    /// that fails after that leaves a key the retry reuses, whereas the
    /// reverse order could put a key on the Host that this device can no
    /// longer decrypt with. Re-registration with the same token is
    /// idempotent — the file keeps one entry per device token.
    @discardableResult
    func register(
        hostID: UUID,
        hostName: String,
        deviceToken: APNSDeviceToken,
        notify: NotificationTriggerPreferences = NotificationTriggerPreferences(),
        relayBaseURL: URL? = nil,
        over transport: any Transport
    ) async throws -> NotificationKeyRecord {
        let record = try hostRecord(hostID: hostID, hostName: hostName)
        try keys.save(record)
        let file = try NotificationRegistrationFile.decode(
            try await transport.readNotificationRegistration())
        let entry = NotificationDeviceEntry(
            token: deviceToken, key: record.key, notify: notify)
        try await transport.replaceNotificationRegistration(
            try file.upserting(entry).encoded())
        if let resolvedRelayURL = NotificationRelayEndpoint.resolve(
            customBaseURL: relayBaseURL)
        {
            try await applyRelayURL(resolvedRelayURL, over: transport)
        }
        return record
    }

    /// Writes the resolved Push Relay base URL into the Host's `notify.json` so
    /// this Host's notify hook POSTs there (#76). Read-merge-write preserves
    /// the plugin's own knobs (`debounce_ms`, `retry_delay_ms`, and future
    /// fields). The production endpoint is written for the empty/default app
    /// setting; self-builders can still supply an explicit custom URL.
    private func applyRelayURL(_ relayBaseURL: URL, over transport: any Transport) async throws {
        let config = try NotificationConfigFile.decode(
            try await transport.readNotificationConfig())
        let updated = config.settingRelayURL(relayBaseURL.absoluteString)
        guard updated != config else { return }
        try await transport.replaceNotificationConfig(try updated.encoded())
    }

    /// Revokes this device on one Host: its entry leaves the registration
    /// file, then the local Notification Key record is dropped — in that
    /// order, so a failed remote removal keeps the key that still-armed
    /// pushes need. Removing a device that was never registered is a no-op.
    func remove(
        hostID: UUID,
        deviceToken: APNSDeviceToken,
        over transport: any Transport
    ) async throws {
        if let data = try await transport.readNotificationRegistration() {
            let file = try NotificationRegistrationFile.decode(data)
            if file.containsDevice(token: deviceToken.hex) {
                try await transport.replaceNotificationRegistration(
                    try file.removing(token: deviceToken.hex).encoded())
            }
        }
        try keys.removeRecord(forHost: hostID)
    }

    /// The Host's key record: the existing key when one is stored (the
    /// service extension must keep decrypting with the key the Host already
    /// holds), refreshed with the current display name; a fresh key
    /// otherwise.
    private func hostRecord(hostID: UUID, hostName: String) throws -> NotificationKeyRecord {
        let key = try keys.record(forHost: hostID)?.key ?? NotificationKeyStore.generateKey()
        return NotificationKeyRecord(hostID: hostID, hostName: hostName, key: key)
    }
}

File: android/app/src/main/java/dev/bybee/heeler/notifications/NotificationRegistrationCeremony.kt
package dev.bybee.heeler.notifications

import android.content.Context
import dev.bybee.heeler.core.crypto.NotificationKeyRecord
import dev.bybee.heeler.core.crypto.NotificationKeyStore
import dev.bybee.heeler.core.transport.Transport
import java.util.UUID
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Per-Host notification enrollment. The local key write always precedes the
 * remote registration replace, so a retry never strands an unreadable Host key.
 */
class NotificationRegistrationCeremony(
    private val keys: NotificationKeyStore,
) {
    suspend fun register(
        hostId: String,
        hostName: String,
        fcmToken: String,
        preferences: NotificationTriggerPreferences = NotificationTriggerPreferences(),
        relayUrl: String = NotificationRelaySettings.DEFAULT_RELAY_URL,
        transport: Transport,
    ): NotificationKeyRecord {
        val id = hostId.asUuid()
        require(fcmToken.isNotBlank()) { "FCM registration token must not be blank." }
        val record = hostRecord(id, hostName)
        keys.save(record)

        val registration = NotificationRegistrationFile.decode(transport.readNotificationRegistration())
        val key = record.key
        try {
            transport.replaceNotificationRegistration(
                registration.upsertingFcm(fcmToken, key, preferences).encoded(),
            )
        } finally {
            key.fill(0)
        }
        writeRelayUrl(relayUrl, transport)
        return record
    }

    suspend fun unregister(
        hostId: String,
        fcmToken: String,
        transport: Transport,
    ) {
        val id = hostId.asUuid()
        require(fcmToken.isNotBlank()) { "FCM registration token must not be blank." }
        val existing = transport.readNotificationRegistration()
        if (existing != null) {
            val registration = NotificationRegistrationFile.decode(existing)
            if (registration.containsDevice(fcmToken)) {
                transport.replaceNotificationRegistration(registration.removing(fcmToken).encoded())
            }
        }
        keys.removeRecord(id)
    }

    suspend fun readPreferences(fcmToken: String, transport: Transport): NotificationTriggerPreferences? =
        NotificationRegistrationFile.decode(transport.readNotificationRegistration()).preferences(fcmToken)

    private fun hostRecord(hostId: UUID, hostName: String): NotificationKeyRecord {
        val old = keys.recordForHost(hostId)
        val key = old?.key ?: NotificationKeyStore.generateKey()
        return try {
            NotificationKeyRecord(hostId, hostName.ifBlank { "Host" }, key)
        } finally {
            key.fill(0)
        }
    }

    private suspend fun writeRelayUrl(relayUrl: String, transport: Transport) {
        val normalized = NotificationRelaySettings.validRelayUrlOrNull(relayUrl)
            ?: throw IllegalArgumentException("Relay URL must be an absolute http(s) URL without a query or fragment.")
        val config = NotificationConfigFile.decode(transport.readNotificationConfig())
        val updated = config.settingRelayUrl(normalized)
        if (updated != config) transport.replaceNotificationConfig(updated.encoded())
    }

    companion object {
        fun create(context: Context): NotificationRegistrationCeremony =
            NotificationRegistrationCeremony(NotificationKeyStore.create(context.applicationContext))
    }
}

private fun String.asUuid(): UUID = try {
    UUID.fromString(this)
} catch (_: IllegalArgumentException) {
    throw IllegalArgumentException("Host id must be a UUID.")
}

/** The Host's schema-free notify.json; only relay_url is owned by the app. */
private class NotificationConfigFile private constructor(
    private val fields: Map<String, kotlinx.serialization.json.JsonElement>,
) {
    fun settingRelayUrl(url: String): NotificationConfigFile = NotificationConfigFile(
        fields + (RELAY_URL_KEY to JsonPrimitive(url.trimEnd('/'))),
    )

    fun encoded(): ByteArray = json.encodeToString(JsonObject.serializer(), JsonObject(fields)).encodeToByteArray()

    override fun equals(other: Any?): Boolean = other is NotificationConfigFile && fields == other.fields
    override fun hashCode(): Int = fields.hashCode()

    companion object {
        private const val RELAY_URL_KEY = "relay_url"
        private val json = Json { ignoreUnknownKeys = true }

        fun decode(bytes: ByteArray?): NotificationConfigFile {
            if (bytes == null) return NotificationConfigFile(emptyMap())
            val decoded = try {
                json.parseToJsonElement(bytes.decodeToString()) as? JsonObject
            } catch (_: SerializationException) {
                null
            } catch (_: IllegalArgumentException) {
                null
            }
            return NotificationConfigFile(decoded ?: emptyMap())
        }
    }
}
